src/routes/subscriptions.rs

```diff
@@ -32,59 +32,126 @@ impl TryFrom<FormData> for NewSubscriber {
     }
 }
 
+pub enum SubscribeError {
+    Validation(String),
+    StoreToken(StoreTokenError),
+    SendEmail(reqwest::Error),
+    Pool(sqlx::Error),
+    InsertSubscriber(sqlx::Error),
+    TransactionCommit(sqlx::Error),
+}
+
+impl std::error::Error for SubscribeError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            SubscribeError::StoreToken(e) => Some(e),
+            SubscribeError::SendEmail(e) => Some(e),
+            SubscribeError::InsertSubscriber(e) => Some(e),
+            SubscribeError::Pool(e) => Some(e),
+            SubscribeError::TransactionCommit(e) => Some(e),
+            SubscribeError::Validation(_) => None,
+        }
+    }
+}
+
+impl std::fmt::Debug for SubscribeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        error_chain_fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for SubscribeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SubscribeError::StoreToken(_) => write!(
+                f,
+                "Failed to store the confirmation token for a new subscriber."
+            ),
+            SubscribeError::SendEmail(_) => write!(f, "Failed to send a confirmation email."),
+            SubscribeError::Validation(e) => write!(f, "{}", e),
+            SubscribeError::InsertSubscriber(_) => {
+                write!(f, "Failed to insert new subscriber in the database.")
+            }
+            SubscribeError::Pool(_) => {
+                write!(f, "Failed to acquire a Postgres connection from the pool.")
+            }
+            SubscribeError::TransactionCommit(_) => write!(
+                f,
+                "Failed to commit SQL transaction to store a new subscriber"
+            ),
+        }
+    }
+}
+
+impl IntoResponse for SubscribeError {
+    fn into_response(self) -> axum::response::Response {
+        match self {
+            SubscribeError::Validation(_) => http::StatusCode::BAD_REQUEST.into_response(),
+            SubscribeError::InsertSubscriber(_)
+            | SubscribeError::Pool(_)
+            | SubscribeError::TransactionCommit(_)
+            | SubscribeError::SendEmail(_)
+            | SubscribeError::StoreToken(_) => {
+                http::StatusCode::INTERNAL_SERVER_ERROR.into_response()
+            }
+        }
+    }
+}
+
+impl From<reqwest::Error> for SubscribeError {
+    fn from(e: reqwest::Error) -> Self {
+        Self::SendEmail(e)
+    }
+}
```

Comment:
All of these `impl From` blocks are necessary for us to take advantage of the `?` operator in the `subscriptions` function